# **Segmenting and Clustering Neighbourhoods in Toronto**

##Importing data

In [1]:
import pandas as pd
import numpy as np
import requests

Get the HTML page of Wiki, and using read_html we convert the html data into list of Data frame objects.


In [2]:


wiki = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_page = requests.get(wiki)

wiki_data = pd.read_html(wiki_page.content, header = 0)[0] #We need the first table alone, so dropping the other tables
wiki_data.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Remove cells which have borrough not assigned

In [3]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code  ...                                      Neighbourhood
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
5           M6A  ...                   Lawrence Manor, Lawrence Heights
6           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
160         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
165         M4Y  ...                               Church and Wellesley
168         M7Y  ...  Business reply mail Processing Centre, South C...
169         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [4]:
df = df.groupby(['Postal Code']).head()
df

Postal Code  ...                                      Neighbourhood
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
5           M6A  ...                   Lawrence Manor, Lawrence Heights
6           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
160         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
165         M4Y  ...                               Church and Wellesley
168         M7Y  ...  Business reply mail Processing Centre, South C...
169         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

In [5]:
df = df.reset_index()
df

index  ...                                      Neighbourhood
0        2  ...                                          Parkwoods
1        3  ...                                   Victoria Village
2        4  ...                          Regent Park, Harbourfront
3        5  ...                   Lawrence Manor, Lawrence Heights
4        6  ...        Queen's Park, Ontario Provincial Government
..     ...  ...                                                ...
98     160  ...      The Kingsway, Montgomery Road, Old Mill North
99     165  ...                               Church and Wellesley
100    168  ...  Business reply mail Processing Centre, South C...
101    169  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102    178  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 4 columns]

In [6]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code  ...                                      Neighbourhood
0           M3A  ...                                          Parkwoods
1           M4A  ...                                   Victoria Village
2           M5A  ...                          Regent Park, Harbourfront
3           M6A  ...                   Lawrence Manor, Lawrence Heights
4           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
98          M8X  ...      The Kingsway, Montgomery Road, Old Mill North
99          M4Y  ...                               Church and Wellesley
100         M7Y  ...  Business reply mail Processing Centre, South C...
101         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

In [7]:
df.head(12)

Postal Code  ...                                      Neighbourhood
0          M3A  ...                                          Parkwoods
1          M4A  ...                                   Victoria Village
2          M5A  ...                          Regent Park, Harbourfront
3          M6A  ...                   Lawrence Manor, Lawrence Heights
4          M7A  ...        Queen's Park, Ontario Provincial Government
5          M9A  ...            Islington Avenue, Humber Valley Village
6          M1B  ...                                     Malvern, Rouge
7          M3B  ...                                          Don Mills
8          M4B  ...                    Parkview Hill, Woodbine Gardens
9          M5B  ...                           Garden District, Ryerson
10         M6B  ...                                          Glencairn
11         M9B  ...  West Deane Park, Princess Gardens, Martin Grov...

[12 rows x 3 columns]

print the number of rows

In [8]:
rows = len(df['Postal Code'].unique())
print("Total number of rows is", rows)

Total number of rows is 103


#**Part 2**

In [9]:
pip install geocoder

     |████████████████████████████████| 102kB 3.4MB/s 


In [10]:
import geocoder

geocoder wasn't working that's why using csv file given in the course

In [11]:
url = 'http://cocl.us/Geospatial_data'

In [12]:
df_geo = pd.read_csv(url)
df_geo.head(12)

Postal Code   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849

In [13]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", df_geo.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [14]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [15]:
df_geo.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [16]:
combined_data = df.join(df_geo.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data.head(12)

Postal Code           Borough  ...   Latitude  Longitude
0          M3A        North York  ...  43.753259 -79.329656
1          M4A        North York  ...  43.725882 -79.315572
2          M5A  Downtown Toronto  ...  43.654260 -79.360636
3          M6A        North York  ...  43.718518 -79.464763
4          M7A  Downtown Toronto  ...  43.662301 -79.389494
5          M9A         Etobicoke  ...  43.667856 -79.532242
6          M1B       Scarborough  ...  43.806686 -79.194353
7          M3B        North York  ...  43.745906 -79.352188
8          M4B         East York  ...  43.706397 -79.309937
9          M5B  Downtown Toronto  ...  43.657162 -79.378937
10         M6B        North York  ...  43.709577 -79.445073
11         M9B         Etobicoke  ...  43.650943 -79.554724

[12 rows x 5 columns]

In [17]:
combined_data.shape

(103, 5)

We get 103 rows as expected when we do a inner join, so we have good data.



---



#**Part 3**

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [18]:
from geopy.geocoders import Nominatim

In [19]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


visualize the map of Toronto

In [20]:
import folium

In [21]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [24]:
CLIENT_ID = 'LXH41KTVMSF1PHL0KPYQI5WV0VKZJHRZDQJX4IYY5OBPQ5SQ' 
CLIENT_SECRET = 'Y4M1UN5ZAGX5WLHISLMHN1N41NYBIV1JWWR5FX1SKH2RKAKM'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: LXH41KTVMSF1PHL0KPYQI5WV0VKZJHRZDQJX4IYY5OBPQ5SQ
CLIENT_SECRET:Y4M1UN5ZAGX5WLHISLMHN1N41NYBIV1JWWR5FX1SKH2RKAKM



Next, we create a function to get all the venue categories in Toronto

In [25]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [26]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [27]:
venues_in_toronto.shape

(1320, 5)

we have 1320 records and 5 columns. Checking sample data

In [28]:
venues_in_toronto.head()

Neighbourhood  ...         Venue Category
0         Parkwoods  ...                   Park
1         Parkwoods  ...      Food & Drink Shop
2  Victoria Village  ...           Hockey Arena
3  Victoria Village  ...  Portuguese Restaurant
4  Victoria Village  ...            Coffee Shop

[5 rows x 5 columns]

Checking the Venues based on Neighbourhood

In [29]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  ...         Venue Category
0                                             Parkwoods  ...                   Park
1                                             Parkwoods  ...      Food & Drink Shop
2                                      Victoria Village  ...           Hockey Arena
3                                      Victoria Village  ...  Portuguese Restaurant
4                                      Victoria Village  ...            Coffee Shop
...                                                 ...  ...                    ...
1304  Mimico NW, The Queensway West, South of Bloor,...  ...            Wings Joint
1305  Mimico NW, The Queensway West, South of Bloor,...  ...           Burger Joint
1306  Mimico NW, The Queensway West, South of Bloor,...  ...         Discount Store
1307  Mimico NW, The Queensway West, South of Bloor,...  ...        Supplement Shop
1308  Mimico NW, The Queensway West, South of Bloor,...  ...                 Bakery

[413 rows x 5 columns]

So there are 413 records for each neighbourhood.

Checking for the maximum venue categories

In [30]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  ...                            Venue
Venue Category                                                         ...                                 
Accessories Store                                   Wexford, Maryvale  ...                      Puffin Gear
Airport                                                     Downsview  ...  Toronto Downsview Airport (YZD)
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...  ...                Billy Bishop Café
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...  ...                           Gate 8
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...  ...                    Porter Lounge
...                                                               ...  ...                              ...
Warehouse Store                                      Thorncliffe Park  ...                           Costco
Wine Bar                                              Studio District  ...                  Paris Paris Bar
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...  ...                       Wingporium
Women's Store                        Lawrence Manor, Lawrence Heights  ...                    Maximum Woman
Yoga Studio                            University of Toronto, Harbord  ...                        YogaSpace

[237 rows x 4 columns]

There are around 237 different types of Venue Categories. Interesting!

#One Hot encoding the venue Categories

In [31]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  ...  Women's Store  Yoga Studio
0                     0        0  ...              0            0
1                     0        0  ...              0            0
2                     0        0  ...              0            0
3                     0        0  ...              0            0
4                     0        0  ...              0            0
...                 ...      ...  ...            ...          ...
1315                  0        0  ...              0            0
1316                  0        0  ...              0            0
1317                  0        0  ...              0            0
1318                  0        0  ...              0            0
1319                  0        0  ...              0            0

[1320 rows x 237 columns]

Adding the neighbourhood to the encoded dataframe

In [32]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  ...  Women's Store  Yoga Studio
0         Parkwoods                  0  ...              0            0
1         Parkwoods                  0  ...              0            0
2  Victoria Village                  0  ...              0            0
3  Victoria Village                  0  ...              0            0
4  Victoria Village                  0  ...              0            0

[5 rows x 238 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [33]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  ...  Yoga Studio
0                                        Agincourt  ...          0.0
1                           Alderwood, Long Branch  ...          0.0
2  Bathurst Manor, Wilson Heights, Downsview North  ...          0.0
3                                  Bayview Village  ...          0.0
4                Bedford Park, Lawrence Manor East  ...          0.0

[5 rows x 238 columns]

Let's make a function to get the top most common venue categories

In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [35]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood  ...       10th Most Common Venue
0                                        Agincourt  ...          Dumpling Restaurant
1                           Alderwood, Long Branch  ...  Eastern European Restaurant
2  Bathurst Manor, Wilson Heights, Downsview North  ...                  Supermarket
3                                  Bayview Village  ...          Dumpling Restaurant
4                Bedford Park, Lawrence Manor East  ...                      Butcher

[5 rows x 11 columns]

Let's make the model to cluster our Neighbourhoods

In [37]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [38]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans


KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

In [39]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 4, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1,
       0, 1, 1, 1, 1, 1, 0, 2], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [40]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [41]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough  ...  9th Most Common Venue  10th Most Common Venue
0         M3A        North York  ...              Drugstore              Donut Shop
1         M4A        North York  ...          Deli / Bodega     Distribution Center
2         M5A  Downtown Toronto  ...         Farmers Market            Dessert Shop
3         M6A        North York  ...  Vietnamese Restaurant        Airport Terminal
4         M7A  Downtown Toronto  ...             Hobby Shop     Fried Chicken Joint

[5 rows x 16 columns]

Drop all the NaN values to prevent data skew

In [42]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [43]:
import matplotlib.cm as cm
import matplotlib.colors as colors

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Make this Notebook Trusted to load map: File -> Trust Notebook

Let's verify each of our clusters

Cluster 1

In [44]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  ...  10th Most Common Venue
0         North York  ...              Donut Shop
21              York  ...              Donut Shop
35         East York  ...               Drugstore
49        North York  ...               Drugstore
61   Central Toronto  ...              Donut Shop
64              York  ...                 Dog Run
66        North York  ...              Donut Shop
85       Scarborough  ...              Donut Shop
91  Downtown Toronto  ...              Donut Shop

[9 rows x 12 columns]

Cluster 2

In [45]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  ...  10th Most Common Venue
1          North York  ...     Distribution Center
2    Downtown Toronto  ...            Dessert Shop
3          North York  ...        Airport Terminal
4    Downtown Toronto  ...     Fried Chicken Joint
7          North York  ...          Discount Store
..                ...  ...                     ...
98          Etobicoke  ...                 Dog Run
99   Downtown Toronto  ...            Burger Joint
100      East Toronto  ...                 Brewery
101         Etobicoke  ...               Drugstore
102         Etobicoke  ...                  Bakery

[88 rows x 12 columns]

Cluster 3

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
45  North York             2.0  ...            Donut Shop                Dog Run

[1 rows x 12 columns]

Cluster 4

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
11  Etobicoke             3.0  ...            Donut Shop                Dog Run

[1 rows x 12 columns]

Cluster 5

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
6  Scarborough             4.0  ...            Donut Shop                Dog Run

[1 rows x 12 columns]


We have successfully cluster Toronto neighbourhood based on venue categories!